### ISSUE TITLE
Internet died during first sync, next sync uploaded the half-downloaded file

### ISSUE THREAD

**Issue by Clef-0** [abraunegg/onedrive#791]
I'm having issues with my broadband at present, and while syncing my OneDrive, my connection died, leaving a PDF file half downloaded. The next sync I attempted continued downloading the other files and then uploaded the previously half-downloaded file. I then reverted the file on the OneDrive online interface. I don't have the terminal log, but the program ended abruptly at something like 60% and returned me back to shell prompt.

**Comment by abraunegg** [comment]
@Clef-0 
Please update your post above with all the required information when raising an issue as per https://github.com/abraunegg/onedrive#reporting-issues

**Comment by abraunegg** [comment]
@Clef-0 
Please can you respond and provide the required details.

**Comment by abraunegg** [comment]
@Clef-0
Please can you respond and provide the required details.

**Comment by Clef-0** [comment]
**Application and Operating System Details:**
*   OS: Linux fedorausb 5.4.17-200.fc31.x86_64 #1 SMP Sat Feb 1 19:00:13 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
*   Are you using a headless system (no gui) or with a gui installed?: I am using GNOME on X11
*   OneDrive Account Type: Personal
*   DMD or LDC compiler version `dmd --version` or `ldmd2 --version`: N/A, using the Fedora repository version
*   Application configuration:
```
onedrive version                    = v2.3.12
Config path                         = /home/niall/.config/onedrive
Config file found in config path    = false
Config option 'check_nosync'        = false
Config option 'sync_dir'            = /home/niall/OneDrive
Config option 'skip_dir'            = 
Config option 'skip_file'           = ~*|.~*|*.tmp
Config option 'skip_dotfiles'       = false
Config option 'skip_symlinks'       = false
Config option 'monitor_interval'    = 45
Config option 'min_notify_changes'  = 5
Config option 'log_dir'             = /var/log/onedrive/
Config option 'sync_root_files'     = false
Selective sync configured           = false
```
*   Curl Version: Output of `curl --version`
```
curl 7.66.0 (x86_64-redhat-linux-gnu) libcurl/7.66.0 OpenSSL/1.1.1d-fips zlib/1.2.11 brotli/1.0.7 libidn2/2.3.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.39.2
Release-Date: 2019-09-11
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz Metalink NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

**Comment by Clef-0** [comment]
This has also been tagged Local Environment Issue which I don't believe to be the case, this issue is in effect a 'feature' request for your app to record which file it is currently downloading so that if the download is interrupted, the incomplete file will not be reuploaded on the next sync. My issue title should have described this better, but I was in a rush.

**Comment by abraunegg** [comment]
@Clef-0 
Any sort of log file you have from the 'event' would b greatly appreciated. This will help try and trace back what actually occurred.

What should be happening is, that an incomplete file download is flagged as just that, incomplete, and then that 'should' be removed, and the file cleanly downloaded next time around - not have the half size / incomplete file uploaded as new or even replacing the existing file causing data loss. 

With the reproduction steps, will see if I can reproduce to diagnose further.

>This has also been tagged Local Environment Issue which I don't believe to be the case

Without any details provided when you first raised the case - no assessment could be made at the time to what is going on. Your issue lacked any depth or substance. This is why there are the templates provided for raises issues which provides prompts for all the needed details to assist with determining what is occurring. I was also on the verge of closing this issue this morning and would also have cited 'lack of response' as the closure reason.

From your updated details and description this sounds like one of the original 'skilion' bugs that was also fixed - so again, without any details when you first raised the case, hard to determine exactly what is going on - no version details provided etc.

To rule this out as a possibility, please can you also check and validate that you do not have any older 'skilion' version lurking anywhere on your system - just in case.

**Comment by abraunegg** [comment]
@Clef-0 
1. Start a manual sync

```
[alex@centos7full onedrive]$ ./onedrive --confdir '~/.config/onedrive-personal/' --synchronize --verbose
Using Config Dir: /home/alex/.config/onedrive-personal/
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.13-14-gfc6e2d9
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5263851520
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Downloading file large_file.data ... 
Downloading  40% |oooooooooooooooo                        |   ETA   00:00:19         
```

2. In another session, block HTTPS traffic

```
firewall-cmd --direct --add-rule ipv4 filter OUTPUT 0 -p tcp -m tcp --dport=443 -j DROP
firewall-cmd --reload
iptables -L -n | grep 443
```

2a. Download process from step 1 above times out, CTRL-C eventually to stop:
```
[alex@centos7full onedrive]$ ./onedrive --confdir '~/.config/onedrive-personal/' --synchronize --verbose
Using Config Dir: /home/alex/.config/onedrive-personal/
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.13-14-gfc6e2d9
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5263851520
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Downloading file large_file.data ... 
Downloading  40% |oooooooooooooooo                        |   ETA   00:00:19 Retry sync count: 1: Timeout was reached on handle 16E5390                                            
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Downloading  45% |oooooooooooooooooo                      |   ETA   00:06:29 ERROR: OneDrive returned an error with the following message:                                         
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
  Error Message: There was a timeout in accessing the Microsoft OneDrive service - Internet connectivity issue?
^C
[alex@centos7full onedrive]$ 


```

3. List all files in OneDrive local directory just to see what is there
```
[alex@centos7full OneDrivePersonal]$ ls -la
total 4
drwxrwxr-x.  2 alex alex    6 Feb 12 17:24 .
drwx------. 36 alex alex 4096 Feb 12 17:12 ..
[alex@centos7full OneDrivePersonal]$ 

```

Nothing downloaded - file that was being downloaded, which was incomplete - is not there ...

4. Unblock HTTPS
```
service firewalld restart
iptables -L -n | grep 443
```


5. Restart the manual sync process
```
[alex@centos7full onedrive]$ ./onedrive --confdir '~/.config/onedrive-personal/' --synchronize --verbose
Using Config Dir: /home/alex/.config/onedrive-personal/
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.13-14-gfc6e2d9
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5263851520
Fetching details for OneDrive Root
OneDrive Root exists in the database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Downloading file large_file.data ... 
Downloading 100% |oooooooooooooooooooooooooooooooooooooooo| DONE IN 00:04:59                                                                                                       
done.
Uploading differences of .
Processing root
The directory has not changed
Processing large_file.data
The file has not changed
Uploading new items of .
Applying changes of Path ID: 66D53BE8A5056ECA!101
[alex@centos7full onedrive]$ 


```

File is downloaded, no partial file is uploaded



In short - cannot replicate said issue ... 

Please can you:

1. Re-create the issue using 'master' or latest full release (v2.3.13)
2. When re-creating issue, ensure debug logging is enabled (`--verbose --verbose`)

**Comment by abraunegg** [comment]
@Clef-0 
Any update in terms of replicating the issue in your environment and providing the required logs?

**Comment by abraunegg** [comment]
@Clef-0 
In trying to replicate this further - if CTRL-C is just used during the download, rather than Internet timeout, the following occurs:

```
Using Config Dir: /home/alex/.config/onedrive-personal/
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.13-14-gfc6e2d9
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5263851520
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Downloading file large_file.data ... 
Downloading  40% |oooooooooooooooo                        |   ETA   00:00:19 ^C                                                                                                               
[alex@centos7full onedrive]$ ./onedrive --confdir '~/.config/onedrive-personal/' --synchronize --verbose 
Using Config Dir: /home/alex/.config/onedrive-personal/
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.3.13-14-gfc6e2d9
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5263851520
Fetching details for OneDrive Root
OneDrive Root exists in the database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
The local item has a different modified time 2020-Feb-14 20:17:08Z remote is 2020-Feb-12 06:15:00Z
The local item has a different hash
Local item modified time is newer based on UTC time conversion - keeping local item
Uploading differences of .
Processing root
The directory has not changed
Processing large_file.data
The file last modified time has changed
The file content has changed
Uploading modified file large_file.data ... 
Uploading  20% |oooooooo                                |   ETA   --:--:--:^C     
```

This sort of looks like your issue - please can you confirm

**Comment by abraunegg** [comment]
@Clef-0 
Closing issue due to the following reasons:
* Lack of response
* Potential issue identified & fixed, PR merged into master
